databricks/koalas/metadata.py

```diff
@@ -30,69 +32,192 @@ class Metadata(object):
     """
     Manages column names and index information.
 
-    :ivar _column_fields: list of the Spark field names to be seen as columns in Koalas DataFrame.
-    :ivar _index_info: list of pair holding the Spark field names for indexes,
+    :ivar _data_columns: list of the Spark field names to be seen as columns in Koalas DataFrame.
+    :ivar _index_pairs: list of pair holding the Spark field names for indexes,
                        and the index name to be seen in Koalas DataFrame.
+
+    .. note:: this is an internal class. It is not supposed to be exposed to users and users
+        should not directly access to it.
+
+    Metadata represents the index information for a DataFrame it belongs to. For instance,
+    if we have a Koalas DataFrame as below, Pandas DataFrame does not store the index as columns.
+
+    >>> kdf = ks.DataFrame({
+    ...     'A': [1, 2, 3, 4],
+    ...     'B': [5, 6, 7, 8],
+    ...     'C': [9, 10, 11, 12],
+    ...     'D': [13, 14, 15, 16],
+    ...     'E':[17, 18, 19, 20]})
+    >>> kdf  # doctest: +NORMALIZE_WHITESPACE
+       A  B   C   D   E
+    0  1  5   9  13  17
+    1  2  6  10  14  18
+    2  3  7  11  15  19
+    3  4  8  12  16  20
+
+    However, all columns including index column are also stored in Spark DataFrame internally
+    as below.
+
+    >>> kdf.to_spark().show()  # doctest: +NORMALIZE_WHITESPACE
+    +-----------------+---+---+---+---+---+
+    |__index_level_0__|  A|  B|  C|  D|  E|
+    +-----------------+---+---+---+---+---+
+    |                0|  1|  5|  9| 13| 17|
+    |                1|  2|  6| 10| 14| 18|
+    |                2|  3|  7| 11| 15| 19|
+    |                3|  4|  8| 12| 16| 20|
+    +-----------------+---+---+---+---+---+
+
+    In order to fill this gap, the current metadata is used by mapping Spark's internal column
+    to Koalas' index. See the method below:
+
+    * `data_columns` represents non-indexing columns
+
+    * `index_columns` represents internal index columns
+
+    * `columns` represents all columns
+
+    * `index_names` represents the external index name
+
+    * `index_pairs` is zipped pairs of `index_columns` and `index_names`
+
+    >>> metadata = kdf._metadata
+    >>> metadata.data_columns
+    ['A', 'B', 'C', 'D', 'E']
+    >>> metadata.index_columns
+    ['__index_level_0__']
+    >>> metadata.columns
+    ['__index_level_0__', 'A', 'B', 'C', 'D', 'E']
+    >>> metadata.index_names
+    [None]
+    >>> metadata.index_pairs
+    [('__index_level_0__', None)]
+
+    In case that index is set to one of the existing column as below:
+
+    >>> kdf1 = kdf.set_index("A")
+    >>> kdf1  # doctest: +NORMALIZE_WHITESPACE
+       B   C   D   E
+    A
+    1  5   9  13  17
+    2  6  10  14  18
+    3  7  11  15  19
+    4  8  12  16  20
+
+    >>> kdf1.to_spark().show()  # doctest: +NORMALIZE_WHITESPACE
```

Comment:
do we just lose the index column when to_spark is called? i think we should keep those. cc @thunterdb

Reply:
oh nvm it's A